### DISCUSSION TITLE
Orchestra on Fit iot lab and z1 motes

### DISCUSSION THREAD

**Discussion by hazelPhd** [contiki-ng/contiki-ng#1890] [Q&A]
Hello ..

I have written an application that is based on two UDP connections, unicast and a broadcast.
With CSMA and TSCH, it works fine with both cooja in a z1 mote and Fit IoT labs m3 nodes.
However, when I try to use the orchestra module I either keep getting not reachable yet for the whole simulation duration, or the simulation will stop at illegal read for the msp430 compiler.
On Fit Iotlab, all I keep getting is a not reachable yet message, the DODAG is never established and the nodes keep on sending a DIS message.

Would this be a platform issue or something in the code?
Any insights on this would be helpful.

Thank you.

**Comment by hazelPhd** [comment]
This is the output that I get from the Fit IotLab 

> 1653900288.450742;m3-63;[INFO: TSCH      ] scanning on channel 25
> 1653900288.496071;m3-36;[INFO: TSCH      ] scanning on channel 26
> 1653900288.862993;m3-82;[INFO: RPL       ] Sending a DIS to ff02::1a
> 1653900288.878872;m3-82;[WARN: 6LoWPAN   ] output: failed to calculate payload size - dropping packet
> 1653900288.920587;m3-15;[INFO: TSCH      ] scanning on channel 25
> 1653900288.924000;m3-9;[INFO: RPL       ] Sending a DIS to ff02::1a
> 1653900288.924276;m3-9;[WARN: 6LoWPAN   ] output: failed to calculate payload size - dropping packet
> 1653900288.998908;m3-31;[INFO: TSCH      ] scanning on channel 15
> 1653900288.999599;m3-33;[INFO: TSCH      ] scanning on channel 26
> 1653900289.001649;m3-66;[INFO: TSCH      ] scanning on channel 15
> 1653900289.032557;m3-34;[INFO: TSCH      ] scanning on channel 20
> 1653900289.042097;m3-27;[INFO: RPL       ] Sending a DIS to ff02::1a
> 1653900289.055365;m3-25;[INFO: RPL       ] Sending a DIS to ff02::1a
> 1653900289.055637;m3-25;[WARN: 6LoWPAN   ] output: failed to calculate payload size - dropping packet
> 1653900289.057930;m3-27;[WARN: 6LoWPAN   ] output: failed to calculate payload size - dropping packet
> 1653900289.078767;m3-79;[INFO: TSCH      ] scanning on channel 15
> 1653900289.081682;m3-57;[INFO: TSCH      ] scanning on channel 20
> 1653900289.102783;m3-82;[INFO: TSCH      ] scanning on channel 15
> 1653900289.114451;m3-2;[INFO: TSCH      ] scanning on channel 15

Any idea why I keep getting that?
Thnaks.

**Comment by atiselsts** [comment]
Contiki-NG with Orchestra used to work on FIT-IoT lab, unless is a new and major problem somewhere in the more recent changes, it should still work.

With Z1 node you perhaps are running out of RAM? What does `msp430-size` say?

**Reply by hazelPhd** [reply]
With z1 I figured this is the case that's why I switched to cooja mote, and I got nice results.  For Orchestra only though, when I try TSCH with the default 6TiSCH on cooja, I get a very low delivery ratio with messages being dropped so often. I tried to change the TSCH link number which enhanced the results for a 5x5 network, but things are still as bad with a 7x7 network. This is the configuration that I played with:

`#define RPL_CONF_MOP RPL_MOP_STORING_MULTICAST
#define TSCH_SCHEDULE_CONF_MAX_LINKS 64
#define QUEUEBUF_CONF_NUM 32
#define NBR_TABLE_CONF_MAX_NEIGHBORS 32
#define NETSTACK_MAX_ROUTE_ENTRIES 64
#define TSCH_SCHEDULE_CONF_MAX_SLOTFRAMES 10` 

In the 5x5, slot number 7 worked fine as well.

This is what I get when I do 7x7 in cooja mote with TSCH 6TiSCH:

> 380485384 19 [INFO: TSCH-LOG  ] {asn 00.00009442 link  0   7   0  0  0 ch 26} uc-1-0 tx LL-0013->LL-001b, len  74, seq 145, st 2  8
> 380554128 18 [INFO: TSCH Pkt  ] ! parse_eb: frame is not a TSCH beacon. Frame version 2, type 1, FCF 41 e9
>380554128 18 [INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0015.0015.0015.0015 to 0xabcd/ffff.0000.0000.0000
>380626216 43 [INFO: TSCH-LOG  ] {asn 00.00009450 link  0   7   0  0  0 ch 15} uc-1-0 tx LL-002b->LL-002f, len 100, seq 105, st 2  7

Any thoughts on that?
Thanks.

**Comment by arurke** [comment]
I have been running Orchestra (Contiki-NG 4.6 though) on FIT IoT-lab lately without issues, with more or less default settings. Those warnings does not look good, are you using security? Or setting some very limited packet buffer sizes?

**Reply by hazelPhd** [reply]
I haven't changed any settings, and no, I'm not using security. I added the orchestra module through the makefile and executed it from there. It is wired as when I use TSCH with 6TiSCH, I get some results; they are not sound, mind you, but the network is formed, and messages are being sent; however, with Orchestra, nothing!

**Reply by hazelPhd** [reply]
Can you please let me know about the sort of application you tested using Orchestra on Fit IoTlab?
or if there are any specific configurations that you've changed?
Thanks.

**Reply by hazelPhd** [reply]
this what I get when I try orchestra on Fit IoT Lab:
`1653900267.147955;m3-41;[INFO: TSCH      ] scanning on channel 20
1653900267.148083;m3-42;[INFO: TSCH      ] scanning on channel 15
1653900267.148247;m3-43;[INFO: TSCH      ] scanning on channel 15
1653900267.148375;m3-44;[INFO: TSCH      ] scanning on channel 26
1653900267.148534;m3-45;���׹{���9��j������v{����������?v{�����q;ӈq��}r�;�[INFO: TSCH      ] scanning on channel 20
1653900267.148697;m3-46;[INFO: TSCH      ] scanning on channel 20
1653900267.148841;m3-48;[INFO: TSCH      ] scanning on channel 15
1653900267.148968;m3-49;[INFO: TSCH      ] scanning on channel 25
1653900267.149098;m3-5;�[INFO: TSCH      ] scanning on channel 26
1653900267.149229;m3-51;[INFO: TSCH      ] scanning on channel 26
1653900267.149356;m3-52;[INFO: TSCH      ] scanning on channel 20
1653900267.149483;m3-53;[INFO: TSCH      ] scanning on channel 15
1653900267.149608;m3-54;[INFO: TSCH      ] scanning on channel 25
1653900267.149735;m3-55;[INFO: TSCH      ] scanning on channel 20
1653900267.149861;m3-56;[INFO: TSCH      ] scanning on channel 26
1653900267.149986;m3-57;[INFO: TSCH      ] scanning on channel 26
1653900267.150114;m3-58;[INFO: TSCH      ] scanning on channel 15
1653900267.150241;m3-59;[INFO: TSCH      ] scanning on channel 26
1653900267.150371;m3-6;�[INFO: TSCH      ] scanning on channel 26
1653900267.150512;m3-61;[INFO: TSCH      ] scanning on channel 26
1653900267.150639;m3-62;[INFO: TSCH      ] scanning on channel 20
1653900267.150779;m3-64;[INFO: TSCH      ] scanning on channel 20
1653900267.150905;m3-65;[INFO: TSCH      ] scanning on channel 15
1653900267.151045;m3-67;[INFO: TSCH      ] scanning on channel 25
1653900267.151172;m3-68;[INFO: TSCH      ] scanning on channel 15
1653900267.151251;m3-69;[INFO: TSCH      ] scanning on channel 20
1653900267.151312;m3-70;[INFO: TSCH      ] scanning on channel 25
1653900267.151393;m3-71;[INFO: TSCH      ] scanning on channel 25
1653900267.151455;m3-72;[INFO: TSCH      ] scanning on channel 20
1653900267.151516;m3-74;[INFO: TSCH      ] scanning on channel 25
1653900267.151576;m3-75;[INFO: TSCH      ] scanning on channel 26
1653900267.151635;m3-76;[INFO: TSCH      ] scanning on channel 25
1653900267.151694;m3-77;[INFO: TSCH      ] scanning on channel 26
1653900267.151802;m3-83;[INFO: TSCH      ] scanning on channel 15
1653900267.152058;m3-60;[INFO: TSCH      ] scanning on channel 15
1653900267.196636;m3-40;[INFO: TSCH      ] scanning on channel 25
1653900267.202607;m3-29;[INFO: TSCH      ] scanning on channel 25
1653900267.202890;m3-47;[INFO: TSCH      ] scanning on channel 26
1653900267.205469;m3-24;[INFO: TSCH      ] scanning on channel 26
1653900267.211597;m3-1;[INFO: RPL       ] Sending a multicast-DIO with rank 128
1653900267.211828;m3-1;[INFO: Frame 15.4] Out:  1 ffff.0000.0000.0000 14 80 (94)
1653900267.211956;m3-1;[DBG : TSCH Queue] packet is added put_index 0, packet 0x20003430
1653900267.212066;m3-1;[INFO: TSCH      ] send packet to ffff.ffff.ffff.ffff with seqno 82, queue 1 1, len 14 94
1653900267.236344;m3-63;[INFO: TSCH      ] scanning on channel 25
1653900267.259552;m3-1;[INFO: TSCH      ] packet sent to 0000.0000.0000.0000, seqno 114, status 0, tx 1
1653900267.259841;m3-1;[INFO: TSCH-LOG  ] {asn 00.00000136 link  2  31   0  0  1 ch 20} bc-1-0 tx LL-2354->LL-NULL, len  94, seq 114, st 0  1
1653900267.260830;m3-3;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
1653900267.260994;m3-3;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
1653900267.262156;m3-72;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
1653900267.262285;m3-72;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
1653900267.262392;m3-72;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
1653900267.262495;m3-72;[DBG : TSCH      ] ! failed to parse EB (len 94)
1653900267.263811;m3-86;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
1653900267.263940;m3-86;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
1653900267.264047;m3-86;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
1653900267.264178;m3-86;[DBG : TSCH      ] ! failed to parse EB (len 94)
1653900267.264591;m3-82;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
1653900267.264704;m3-82;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
1653900267.264807;m3-82;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
1653900267.264907;m3-82;[DBG : TSCH      ] ! failed to parse EB (len 94)
1653900267.276882;m3-3;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
1653900267.277139;m3-3;[DBG : TSCH      ] ! failed to parse EB (len 94)`

**Reply by arurke** [reply]
Sorry I don't have time to look at it now, but I have an ambition to do some tests on the lab this weekend - I will try to do a vanilla orchestra one as well. But as Atis mentioned, I would be surprised if Orchestra is not working at all. Have you tried checking out v4.6?

My application was a simple data collection - all nodes sending to root. Some quick excerpts:

Makefile:
```
MAKE_MAC = MAKE_MAC_TSCH
MAKE_ROUTING = MAKE_ROUTING_RPL_CLASSIC
MODULES += os/services/orchestra
```

project-conf.h:
```
#define ORCHESTRA_CONF_UNICAST_MAX_CHANNEL_OFFSET 2
#define ORCHESTRA_CONF_COLLISION_FREE_HASH    1
#define ORCHESTRA_CONF_MAX_HASH               49
#define ORCHESTRA_CONF_UNICAST_SENDER_BASED   1
#define TSCH_CONF_BURST_MAX_LEN           0
```

**Reply by hazelPhd** [reply]
Thank you for your reply.

I was using the same configration and did a new run with the v4.6 but still the same, this a sample of the output:

> 1654163573.242889;m3-3;[INFO: TSCH      ] scanning on channel 20
> 1654163573.245486;m3-1;[INFO: RPL       ] Sending a multicast-DIO with rank 128
> 1654163573.245735;m3-1;[INFO: TSCH      ] send packet to ffff.ffff.ffff.ffff with seqno 82, queue 1 1, len 14 94
> 1654163573.405157;m3-1;[INFO: TSCH      ] packet sent to 0000.0000.0000.0000, seqno 115, status 0, tx 1
> 1654163573.408545;m3-76;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.408789;m3-76;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.413335;m3-7;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.413621;m3-7;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.413753;m3-7;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.414655;m3-74;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.414821;m3-74;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.414934;m3-74;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.415430;m3-9;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.415556;m3-9;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.415661;m3-9;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.417338;m3-83;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.417486;m3-83;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.417594;m3-83;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.418033;m3-80;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.418152;m3-80;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.418317;m3-80;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.418704;m3-3;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.418814;m3-3;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.418917;m3-3;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.419367;m3-8;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.419541;m3-8;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.419658;m3-8;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.420052;m3-11;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.420199;m3-11;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.420302;m3-11;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.420990;m3-78;[INFO: TSCH      ] scan: received packet (94 bytes) on channel 20
> 1654163573.421133;m3-78;[INFO: TSCH Pkt  ] ! parse_eb: frame is not a valid TSCH beacon. Frame version 2, type 1, FCF 41 e9
> 1654163573.421239;m3-78;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.424440;m3-76;[INFO: TSCH Pkt  ] ! parse_eb: frame was from 0xabcd/0200.0000.0000.2354 to 0xabcd/ffff.0000.0000.0000
> 1654163573.631224;m3-77;[INFO: TSCH      ] scanning on channel 15
> 1654163573.785197;m3-32;[INFO: TSCH      ] scanning on channel 20
> 1654163573.807599;m3-25;[INFO: TSCH      ] scanning on channel 15
> 1654163573.846104;m3-58;[INFO: TSCH      ] scanning on channel 26

**Reply by arurke** [reply]
EDIT: Switched to different nodes due to faulty node nearby, and corrected project-conf.h

I just ran a simple sanity check:
* Develop branch (ccf2f)
* Grenoble, nodes m3-69, m3-68 and m3-67
* 69 is root node, 68 and 67 sends periodically (5 s) to root

Results: Both nodes associates after about 18 s, and I get 100 % PDR on a 5 minute run.

-----
Relevant snips imo:

Makefile:
```
MAKE_MAC = MAKE_MAC_TSCH
MAKE_ROUTING = MAKE_ROUTING_RPL_CLASSIC
MODULES += os/services/orchestra
```

project-conf.h:
```
#define ORCHESTRA_CONF_RULES { &eb_per_time_source, \
                             &unicast_per_neighbor_rpl_storing, \
                             &default_common }
```

node.c:
```
if(node_id == 0x9682) { // Grenoble m3-69
    NETSTACK_ROUTING.root_start();
    simple_udp_register(&udp_conn, UDP_SERVER_PORT, NULL,
                        UDP_CLIENT_PORT, udp_rx_callback);
}
```

**Reply by hazelPhd** [reply]
When I used v4.6 with RPL Lite, it worked.
However, when I try to tested with RPL Classic I keep getting this error:
`../../os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c: In function 'add_uc_link':
../../os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c:121:21: error: 'local_channel_offset' undeclared (first use in this function)
           timeslot, local_channel_offset);
                     ^
../../os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c:121:21: note: each undeclared identifier is reported only once for each function it appears in
../../os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c: At top level:
cc1: error: unrecognized command line option "-Wno-unused-const-variable" [-Werror]
cc1: all warnings being treated as errors
make: *** [../../../arch/platform/iotlab/Makefile.iotlab:167: build/iotlab/m3/obj/orchestra-rule-unicast-per-neighbor-rpl-storing.o] Error 1`

This is weird since I haven't changed anything, and from the look of things the code should work!

Thank you for your help.

**Reply by arurke** [reply]
I remember there being a type-o with local_cha**nnn**el_offset at some point. If you are unable to debug I would suggest going to v4.7 or develop, and verify you can reproduce my sanity check.

**Reply by hazelPhd** [reply]
I couldn't fix this issue at the moment but I will try to see what is the issue.
However, I have a quick question, when I use orchestra, I get an almost perfect delivery ratio in terms o the messages that are sent to the sink or the root. However, I have a clustering app and the messages that are unicasted to the cluster heads have a much lower delivery ratio, Any idea if this is related to the scheduling or is it about the way RPL downward routing interacts with orchestra?

**Reply by arurke** [reply]
I guess this discussion was continued on gitter: https://gitter.im/contiki-ng/Developers?at=62bdbe03d3c8894f71de060b
